server/webhook.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	cloud "github.com/mattermost/mattermost-cloud/model"
+)
+
+func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
+	payload, err := cloud.WebhookPayloadFromReader(r.Body)
+	if err != nil {
+		p.API.LogError(err.Error())
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	go p.processWebhookEvent(payload)
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (p *Plugin) processWebhookEvent(payload *cloud.WebhookPayload) {
+	str, _ := payload.ToJSON()
+	p.API.LogDebug(str)
+
+	if payload.Type != "installation" {
+		return
+	}
+
+	if payload.NewState != cloud.InstallationStateStable {
+		return
+	}
+
+	install, err := p.getInstallation(payload.ID)
+	if err != nil {
+		p.API.LogError(err.Error(), "installation", install.Name)
+		return
+	}
+	if install == nil {
+		return
+	}
+
+	var installation *cloud.Installation
+
+	switch payload.OldState {
+	case cloud.InstallationStateUpgradeRequested,
+		cloud.InstallationStateUpgradeInProgress,
+		cloud.InstallationStateUpgradeFailed:
+		installation, err = p.cloudClient.GetInstallation(payload.ID)
+		if err != nil {
+			p.API.LogError(err.Error(), "installation", install.Name)
+			return
+		}
+		if installation == nil {
+			p.API.LogError(fmt.Sprintf("could not find installation %s", install.ID))
+		}
+
+		install.Installation = *installation
+
+		message := fmt.Sprintf(`
+Installation %s has been upgraded!
+
+Installation details:
+%s
+`, install.Name, jsonCodeBlock(install.ToPrettyJSON()))
+
+		p.PostBotDM(install.OwnerID, message)
+	case cloud.InstallationStateCreationRequested,
+		cloud.InstallationStateCreationDNS,
+		cloud.InstallationStateCreationFailed:
+		err = p.setupInstallation(install)
+		if err != nil {
+			p.API.LogError(err.Error(), "installation", install.Name)
+			return
+		}
+
+		install.License = "hidden"
+
+		message := fmt.Sprintf(`
+Installation %s is ready!
+
+Access at: https://%s
+
+Login with:
+
+| Username | Password |
+| -- | -- |
+| %s | %s |
+
+Installation details:
+%s
+`, install.Name, install.DNS, defaultAdminUsername, defaultAdminPassword, jsonCodeBlock(install.ToPrettyJSON()))
+
+		p.PostBotDM(install.OwnerID, message)
+	}
+}
+
+func (p *Plugin) handleProfileImage(w http.ResponseWriter, r *http.Request) {
```

Comment:
Copying comment from Lev:

nit: not sure webhook.go#L100 is needed for static assets any longer